examples/fabric-admin/device_manager/DeviceManager.cpp

```diff
@@ -0,0 +1,226 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "DeviceManager.h"
+
+#include <commands/interactive/InteractiveCommands.h>
+
+#include <cstdio>
+#include <string>
+
+using namespace chip;
+using namespace chip::app::Clusters;
+
+namespace {
+
+// Constants
+constexpr uint32_t kSyncPrepareTimeMs = 1000;
+
+} // namespace
+
+// Define the static member
+DeviceManager DeviceManager::sInstance;
+
+void DeviceManager::Init()
+{
+    // TODO: Init mLastUsedNodeId from chip config file
+    mLastUsedNodeId = 1;
+}
+
+NodeId DeviceManager::GetNextAvailableNodeId()
+{
+    mLastUsedNodeId++;
+    VerifyOrDieWithMsg(mLastUsedNodeId < std::numeric_limits<chip::NodeId>::max(), NotSpecified, "No more available NodeIds.");
+
+    return mLastUsedNodeId;
+}
+
+void DeviceManager::UpdateLastUsedNodeId(chip::NodeId nodeId)
+{
+    if (nodeId > mLastUsedNodeId)
+    {
+        ChipLogProgress(NotSpecified, "Updating last used NodeId to " ChipLogFormatX64, ChipLogValueX64(nodeId));
+        mLastUsedNodeId = nodeId;
+    }
+}
+
+void DeviceManager::AddSyncedDevice(const Device & device)
+{
+    mSyncedDevices.insert(device);
+    ChipLogProgress(NotSpecified, "Added synced device: NodeId:" ChipLogFormatX64 ", EndpointId %u",
+                    ChipLogValueX64(device.GetNodeId()), device.GetEndpointId());
+}
+
+Device * DeviceManager::FindDeviceByEndpoint(EndpointId endpointId)
+{
+    for (auto & device : mSyncedDevices)
+    {
+        if (device.GetEndpointId() == endpointId)
+        {
+            return const_cast<Device *>(&device);
```

Comment:
Why are these const_cast needed? mSyncedDevices holds `Device` not `const Device

Reply:
because std::set::iterator returns a constant iterator when used with a set of const elements.